posts/2020-12-14-spring-boot-rental-notifier-bot.adoc

```diff
@@ -0,0 +1,84 @@
+
+== What technologies were used?
+
+I used following technologies:
+
+* Spring boot 2 - Beloved framework by many Java developers for web development
+* MySql database
+* Spring data + Hibernate for data access and manipulation
+* ActiveMQ for communication between services
+* JSoup for HTML parsing
+* Junit for testing
+* Tomcat server
+
+== How does this project work?
+The service periodically fetches and parses the last N pages of a chosen website (what N should be is decided empirically). Then it compares this result with the existing records, deciding whether it is a new offer or updated and keeps history in that case (it could be useful for analytics). Then it passes the records that were changed to the message queue, where they are received by the bot listener. This bot listener triggers Rental Bot abstraction that checks for subscribed users and their preferences. Finally, it filters and sends these recommendations to each user based on their preferences.
+Besides having an Admin user, I wanted to have endpoints to manually fetch or insert data through a REST API.
+
+== What about the architecture and the implementation?
+
+image::../media/2020-12-14-spring-boot-rental-notifier-bot/context-diagram.png[Context Diagram, 600]
+
+Currently the user interacts with the application through the Telegram messenger. Another external dependency is the data source (Pararuis) where the data is extracted from.
+
+And back to the technologies. My intention was to create an extensible server application that could be used for convenient real estate related notifications with possible analytic usage. Here is the component diagram for Rental Notifier:
+
+image::../media/2020-12-14-spring-boot-rental-notifier-bot/component-diagram.png[Component Diagram, 700]
+
+As I mentioned before, I aimed to design the application to be extensible for future needs.
+
+I used a modular approach, creating 3 base modules (packages) - “bot”, “common”, “rental”. In case of an (unlikely) high growth of the application, such structure will allow splitting the monolith into nice simple microservices. As an additional benefit, the modules are easy to navigate, easy to extend with small risk to affect another module. Furthermore, the communication between them is done through the message broker - ActiveMQ. It facilitates adding independent, easy-to-scale and easy-to-extend modules by adding new topics and new listeners.
+As you can see, there are a few components. Following the best practices I tend to do them small, loosely-coupled, doing one thing at the time, open for extensions. As well I used abstractions and interfaces for easy updates - i.e. supporting new data sources (apart from Pararius) would be a question of few new classes implementing existing interfaces.
```

Comment:
Sometimes in English "there are a few" can be confusing :) . Some people might interpret "there are a few components" to mean that "there are many components". But if I understand this paragraph you are trying to say that "there are NOT many components".

So you either need to say:
- "there are few components" (so delete the "a"), or explicitly say
- "there are not many components"